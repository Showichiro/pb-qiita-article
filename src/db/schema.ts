import { relations, text, integer, sqliteTable } from "@/lib";

export const articles = sqliteTable("articles", {
  id: text("id").primaryKey(),
  title: text("title").notNull(),
  userId: text("user_id").notNull(),
  userName: text("user_name").notNull(),
  createdAt: text("created_at").notNull(),
  likesCount: integer("likes_count").notNull(),
  stocksCount: integer("stocks_count").notNull(),
});

export const articleRelation = relations(articles, ({ many }) => ({
  tags: many(tags),
}));

export const tags = sqliteTable("tags", {
  articleId: text("article_id").references(() => articles.id, {
    onDelete: "cascade",
    onUpdate: "cascade",
  }),
  id: integer("id").primaryKey(),
  name: text("name").notNull(),
});

export const tagRelation = relations(tags, ({ one }) => ({
  article: one(articles, {
    fields: [tags.articleId],
    references: [articles.id],
  }),
}));
