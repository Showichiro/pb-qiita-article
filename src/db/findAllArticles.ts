import { DrizzleD1Database } from "@/lib";
import * as schema from "./schema";
import { Article } from "@/schemas";

/**
 * Find all articles.
 *
 * @param db - Database instance.
 * @param config - Configuration object.
 * @returns Array of articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
type FindAllArticlesReturnType = Array<Article>;

/**
 * Configuration object.
 *
 * @param limit - Limit number of articles.
 * @param offset - Offset number of articles.
 * @param sinse - Date since articles.
 * @param until - Date until articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
export type FindAllArticlesConfig = {
  limit: number | null;
  offset: number | null;
  sinse: string | null;
  until: string | null;
};

/**
 * Find all articles.
 *
 * @param db - Database instance.
 * @param config - Configuration object.
 * @returns Array of articles.
 * @example
 * const articles = await findAllArticles(db, { limit: 10, offset: 0 });
 */
export const findAllArticles = async (
  db: DrizzleD1Database<typeof schema>,
  config: FindAllArticlesConfig
): Promise<FindAllArticlesReturnType> => {
  const defaultLimit = 10;
  const defaultOffset = 0;
  const { limit, offset, sinse, until } = config;
  const results = await db.query.articles.findMany({
    limit: limit ?? defaultLimit,
    offset: offset ?? defaultOffset,
    where:
      !sinse && !until
        ? undefined
        : (fileds, { between, gte, lte }) => {
            if (sinse && until) {
              return between(fileds.createdAt, sinse, until);
            }
            if (sinse) {
              return gte(fileds.createdAt, sinse);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          },
    orderBy: (fileds, { desc }) => [desc(fileds.createdAt)],
    with: {
      tags: {
        columns: {
          articleId: false,
          id: false,
          name: true,
        },
      },
    },
  });
  return results;
};
