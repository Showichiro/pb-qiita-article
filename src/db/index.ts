export * as schema from "./schema";
export * from "./findAllArticles";
export * from "./getArticleCountGroupByUser";
export * from "./getLikesCountGroupByUser";
