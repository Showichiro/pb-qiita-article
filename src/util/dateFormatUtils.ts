/**
 * @description Convert a Date object to a string in ISO 8601 format.
 * @param {Date} date The Date object to convert.
 * @returns {string} The ISO 8601 string representation of the Date object.
 * @example
 * const date = new Date();
 * const isoString = dateToDatetimeString(date);
 */
export const dateToDatetimeString = (date: Date): string => {
  return date.toISOString();
};

/**
 * @description Convert a datetime string to a date string.
 * @param {string} datetime The datetime string to convert.
 * @returns {string} The date string representation of the datetime string.
 * @example
 * const datetime = "2022-01-01T12:00:00.000Z";
 * const dateString = dateTimetoDateString(datetime);
 */
export const dateTimetoDateString = (datetime: string): string => {
  return new Date(datetime).toISOString().split("T")[0];
};
