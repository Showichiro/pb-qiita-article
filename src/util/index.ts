export * from "./typeUtils";
export * from "./testUtils";
export * from "./dateFormatUtils";
export * from "./factory";
export * from "./dateParamsUtils";
