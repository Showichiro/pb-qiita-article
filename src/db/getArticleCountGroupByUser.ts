import { DrizzleD1Database, asc, count, desc, between, gte, lte } from "@/lib";
import { schema } from "@/db";
import { ArticleCountGroupByUser } from "@/schemas";

/**
 * Config for getArticleCountGroupByUser.
 * @property {string | null} sinse Start date.
 * @property {string | null} until End date.
 * @property {"asc" | "desc"} sort Sort order. Default is "desc".
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: "2024-01-01T00:00:00Z",
 *   until: "2023-01-01T00:00:00Z",
 *   sort: "asc",
 * };
 * ```
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: null,
 *   until: null,
 *   sort: "asc",
 * };
 * ```
 */
export type RankingConfig = {
  sinse: string | null;
  until: string | null;
  sort?: "asc" | "desc";
};

/**
 * Get article count group by user.
 * @param db DrizzleD1Database instance.
 * @param config Config for getArticleCountGroupByUser.
 * @returns Array of ArticleCountGroupByUser.
 * @example
 * ```ts
 * const config: Config = {
 *   sinse: "2024-01-01T00:00:00Z",
 *   until: "2023-01-01T00:00:00Z",
 *   sort: "asc",
 * };
 * const results = await getArticleCountGroupByUser(db, config);
 * ```
 */
export const getArticleCountGroupByUser = async (
  db: DrizzleD1Database<typeof schema>,
  { sinse, until, sort = "desc" }: RankingConfig
): Promise<Array<ArticleCountGroupByUser>> => {
  const results = await db
    .select({
      count: count(schema.articles.id),
      userId: schema.articles.userId,
      userName: schema.articles.userName,
      createdAt: schema.articles.createdAt,
    })
    .from(schema.articles)
    .where(
      !sinse && !until
        ? undefined
        : (fileds) => {
            if (sinse && until) {
              return between(fileds.createdAt, sinse, until);
            }
            if (sinse) {
              return gte(fileds.createdAt, sinse);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          }
    )
    .groupBy((fileds) => fileds.userId)
    .orderBy((fileds) => [
      sort === "asc" ? asc(fileds.count) : desc(fileds.count),
    ])
    .all();
  return results.map(({ createdAt, ...rest }) => ({ ...rest }));
};
