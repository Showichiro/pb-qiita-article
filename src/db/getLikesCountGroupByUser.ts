import { asc, between, desc, type DrizzleD1Database, gte, lte, sum } from "@/lib";
import type { LikesCountSchema } from "@/schemas";
import { type RankingConfig, schema } from "@/db";

/**
 * Get likes count group by user
 * @param db
 * @param config
 * @returns Array<LikesCountSchema>
 * @example
 * getLikesCountGroupByUser(db, {
 *   since: "2024-01-01T00:00:00Z",
 *   until: null,
 *   sort: "asc",
 * });
 */
export const getLikesCountGroupByUser = async (
  db: DrizzleD1Database<typeof schema>,
  { since, until, sort = "desc" }: RankingConfig,
): Promise<Array<LikesCountSchema>> => {
  const results = await db
    .select({
      totalLikesCount: sum(schema.articles.likesCount),
      userId: schema.articles.userId,
      userName: schema.articles.userName,
      createdAt: schema.articles.createdAt,
    })
    .from(schema.articles)
    .where(
      !since && !until
        ? undefined
        : (fileds) => {
            if (since && until) {
              return between(fileds.createdAt, since, until);
            }
            if (since) {
              return gte(fileds.createdAt, since);
            }
            if (until) {
              return lte(fileds.createdAt, until);
            }
          },
    )
    .groupBy((fileds) => fileds.userId)
    .orderBy((fileds) => [
      sort === "asc"
        ? asc(fileds.totalLikesCount)
        : desc(fileds.totalLikesCount),
    ])
    .all();
  return results.map(({ createdAt, ...rest }) => ({
    ...rest,
  }));
};
