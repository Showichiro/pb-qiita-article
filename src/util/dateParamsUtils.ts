import { dateToDatetimeString } from "./dateFormatUtils";

export const processDateParam = (
  date: string | Date | null | undefined,
): string | null => {
  if (typeof date === "string") return date;
  if (date == null) return null;
  return dateToDatetimeString(date);
};
