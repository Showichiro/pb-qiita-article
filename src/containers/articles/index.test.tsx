import { renderer } from "@/util";
import { Miniflare } from "miniflare";
import { ArticlesContainer } from ".";
import { drizzle } from "drizzle-orm/d1";
import { schema } from "@/db";

describe("ArticlesContainer", async () => {
  const mf = new Miniflare({
    modules: true,
    d1Databases: ["DB"],
    script: `addEventListener("fetch", (event) => {
          event.respondWith(new Response("Hello Miniflare!"));
        })`,
  });

  const db = await mf.getD1Database("DB");

  const record = 10;
  beforeAll(async () => {
    await db.exec(
      "CREATE TABLE `articles` (`id` text PRIMARY KEY NOT NULL,`title` text NOT NULL,`user_id` text NOT NULL,`user_name` text NOT NULL,`created_at` text NOT NULL,`likes_count` integer NOT NULL,`stocks_count` integer NOT NULL);",
    );
    await db.exec(
      "CREATE TABLE `tags` (`article_id` text,`id` integer PRIMARY KEY NOT NULL,`name` text NOT NULL,FOREIGN KEY (`article_id`) REFERENCES `articles`(`id`) ON UPDATE cascade ON DELETE cascade);",
    );
    const promises = [...Array(record)].map(async (_, index) => {
      return await db
        .prepare(
          "INSERT INTO `articles` (`id`, `title`, `user_id`, `user_name`, `created_at`, `likes_count`, `stocks_count`) VALUES (?, ?, ?, ?, ?, ?, ?);",
        )
        .bind(
          `${index}`,
          `title-${index}`,
          `user-${index % 2 === 0 ? 0 : index}`,
          `user-${index % 2 === 0 ? 0 : index}`,
          new Date(index).toISOString(),
          0,
          0,
        )
        .run()
        .then(async () => {
          return await db
            .prepare("INSERT INTO `tags` (`article_id`, `name`) VALUES (?, ?);")
            .bind(`${index}`, `tag-${index}`)
            .run();
        });
    });
    await Promise.all(promises);
  });

  afterAll(async () => {
    await mf.dispose();
  });

  const instance = drizzle(db, { schema, logger: true });

  it("should render article table", async () => {
    const { text } = await renderer(
      <ArticlesContainer
        db={instance}
        config={{ limit: null, offset: null, since: null, until: null }}
      />,
    );
    expect(text).toMatchSnapshot();
  });
});
