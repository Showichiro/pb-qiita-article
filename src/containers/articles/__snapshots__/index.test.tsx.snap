// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticlesContainer > should render article table 1`] = `"<table class="table overflow-x-auto"><thead><tr><th>タイトル</th><th>執筆者</th><th>タグ</th><th>いいね数</th><th>ストック数</th></tr></thead><tbody><tr class="hover"><td><a class="link" href="https://qiita.com/user-9/items/9" target="_blank" rel="noopener noreferrer">title-9</a></td><td><a class="link" href="https://qiita.com/user-9" target="_blank" rel="noopener noreferrer">user-9<span>(user-9)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-9" target="_blank" rel="noopener noreferrer">tag-9</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/8" target="_blank" rel="noopener noreferrer">title-8</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-8" target="_blank" rel="noopener noreferrer">tag-8</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-7/items/7" target="_blank" rel="noopener noreferrer">title-7</a></td><td><a class="link" href="https://qiita.com/user-7" target="_blank" rel="noopener noreferrer">user-7<span>(user-7)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-7" target="_blank" rel="noopener noreferrer">tag-7</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/6" target="_blank" rel="noopener noreferrer">title-6</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-6" target="_blank" rel="noopener noreferrer">tag-6</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-5/items/5" target="_blank" rel="noopener noreferrer">title-5</a></td><td><a class="link" href="https://qiita.com/user-5" target="_blank" rel="noopener noreferrer">user-5<span>(user-5)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-5" target="_blank" rel="noopener noreferrer">tag-5</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/4" target="_blank" rel="noopener noreferrer">title-4</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-4" target="_blank" rel="noopener noreferrer">tag-4</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-3/items/3" target="_blank" rel="noopener noreferrer">title-3</a></td><td><a class="link" href="https://qiita.com/user-3" target="_blank" rel="noopener noreferrer">user-3<span>(user-3)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-3" target="_blank" rel="noopener noreferrer">tag-3</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/2" target="_blank" rel="noopener noreferrer">title-2</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-2" target="_blank" rel="noopener noreferrer">tag-2</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-1/items/1" target="_blank" rel="noopener noreferrer">title-1</a></td><td><a class="link" href="https://qiita.com/user-1" target="_blank" rel="noopener noreferrer">user-1<span>(user-1)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-1" target="_blank" rel="noopener noreferrer">tag-1</a></li></ul></td><td>0</td><td>0</td></tr><tr class="hover"><td><a class="link" href="https://qiita.com/user-0/items/0" target="_blank" rel="noopener noreferrer">title-0</a></td><td><a class="link" href="https://qiita.com/user-0" target="_blank" rel="noopener noreferrer">user-0<span>(user-0)</span></a></td><td><ul><li class="badge badge-lg"><a href="https://qiita.com/tags/tag-0" target="_blank" rel="noopener noreferrer">tag-0</a></li></ul></td><td>0</td><td>0</td></tr></tbody></table>"`;
